File: imageslider/src/main/java/technolifestyle/com/imageslider/CircularFlipperHandler.java
package technolifestyle.com.imageslider;

import android.support.v4.view.ViewPager;

class CircularFlipperHandler implements ViewPager.OnPageChangeListener, FlipperLayout.CurrentPageListener {
    private ViewPager   mViewPager;
    private int         mCurrentPosition;
    private int         mScrollState;

    private FlipperLayout flipperLayout;

    CircularFlipperHandler(final ViewPager viewPager) {
        mViewPager = viewPager;
    }

    @Override
    public void onPageSelected(final int position) {
        mCurrentPosition = position;
        onCurrentPageChanged(mCurrentPosition);
    }

    @Override
    public void onPageScrollStateChanged(final int state) {
        handleScrollState(state);
        mScrollState = state;
    }

    private void handleScrollState(final int state) {
        if (state == ViewPager.SCROLL_STATE_IDLE) {
            setNextItemIfNeeded();
        }
    }

    private void setNextItemIfNeeded() {
        if (!isScrollStateSettling()) {
            handleSetNextItem();
        }
    }

    private boolean isScrollStateSettling() {
        return mScrollState == ViewPager.SCROLL_STATE_SETTLING;
    }

    private void handleSetNextItem() {
        final int lastPosition = mViewPager.getAdapter().getCount() - 1;
        if(mCurrentPosition == 0) {
            mViewPager.setCurrentItem(lastPosition, false);
        } else if(mCurrentPosition == lastPosition) {
            mViewPager.setCurrentItem(0, false);
        }
    }

    @Override
    public void onPageScrolled(final int position, final float positionOffset, final int positionOffsetPixels) {
    }

    @Override
    public int onCurrentPageChanged(int currentPosition) {
        flipperLayout.setCurrentPage(currentPosition);
        return currentPosition;
    }

    @Override
    public void setOnCurrentPageListener(FlipperLayout flipperLayout) {
        this.flipperLayout = flipperLayout;
    }
}

File: imageslider/src/main/java/technolifestyle/com/imageslider/CircularFlipperHandler.kt
package technolifestyle.com.imageslider

import androidx.viewpager.widget.ViewPager
import androidx.viewpager.widget.ViewPager.SCROLL_STATE_DRAGGING
import androidx.viewpager.widget.ViewPager.SCROLL_STATE_IDLE

internal class CircularFlipperHandler(private val mViewPager: ViewPager) : ViewPager.OnPageChangeListener {

    private var scrollState = SCROLL_STATE_IDLE
    private var nextPageAfterScrollStateChanged: Int? = null
    private var currentPageListener: CurrentPageListener? = null


    fun setCurrentPageListener(currentPageListener: CurrentPageListener) {
        this.currentPageListener = currentPageListener
    }

    override fun onPageSelected(position: Int) {
        currentPageListener!!.onCurrentPageChanged(position)
    }

    override fun onPageScrollStateChanged(state: Int) {
        scrollState = state

        nextPageAfterScrollStateChanged?.let {
            mViewPager.currentItem = it
            nextPageAfterScrollStateChanged = null
        }
    }

    override fun onPageScrolled(position: Int, positionOffset: Float, positionOffsetPixels: Int) {
        val count = mViewPager.adapter!!.count

        if (scrollState == SCROLL_STATE_DRAGGING && positionOffset == 0f) {
            if (position == 0) {
                nextPageAfterScrollStateChanged = count - 1
            } else if (position == count - 1) {
                nextPageAfterScrollStateChanged = 0
            }
        }
    }

    internal interface CurrentPageListener {
        fun onCurrentPageChanged(currentPosition: Int)
    }
}
